test, refactor: Embedded ASMap [1/3]: Selected minor preparatory work

fjahr opened the pull request:
This contains some commits from #28792 that can be easily reviewed and merged independently. I hope splitting this change off can make this part move a bit faster and reduce frequency of needed rebases for #28792.

The commits in order:
- Add additional unit test vectors to the asmap interpreter (written by sipa). This helps to ensure that the further refactors in #28792 don't change behavior. 
- Modernizes the logging in `util/asmap.cpp`, I added this while touching the rest of the file all over anyway.
- Adds an `AutoFile::size` helper function with some additional test coverage in a separate commit

fjahr committed:
refactor: Add AutoFile::size

fjahr committed:
refactor: Modernize logging in util/asmap.cpp

sipa committed:
tests: add unit test vectors for asmap interpreter

nervana21 commented:
tACK [942d95d](https://github.com/bitcoin/bitcoin/pull/33026/commits/942d95d3c6c026d0e10a682c18e3a907a6241d8a)

I thoroughly reviewed and understood all code changes. I ran the tests locally and they all passed.

fjahr commented on src/common/args.cpp:
```diff
@@ -273,7 +273,7 @@ fs::path ArgsManager::GetPathArg(std::string arg, const fs::path& default_value)
     if (IsArgNegated(arg)) return fs::path{};
     std::string path_str = GetArg(arg, "");
-    if (path_str.empty()) return default_value;
+    if (path_str.empty() || path_str == "1") return default_value;
```
I can drop this is other reviewers agree that it's confusing. I personally think this is the more intuitive behavior for path args independently of asmap because our bool args all work with both `-foo` and `-foo=1` afaik. If the user wants to use `-asmap` (or any path arg) as a bool then I think they would expect it to work the same as the other bool params. Fwiw, I discovered this because I wrote a functional test that covered `-asmap=1` and I actually expected it work, only after it didn't I added this.

sipa reviewed (commented):
utACK 734737b5930df7cebab83cf0dbe5fd390143f2be

sipa reviewed (commented):
ACK 5bb456c92bbd27ebbba1773832b051968f162b8a

hodlinator commented on src/test/netbase_tests.cpp:
```diff
@@ -610,4 +611,55 @@ BOOST_AUTO_TEST_CASE(isbadport)
+BOOST_AUTO_TEST_CASE(asmap_test_vectors)
+{
+    // Randomly generated encoded ASMap with 128 ranges, up to 20-bit AS numbers.
+    static const auto ASMAP_DATA =
```
nit in 5a63a5a1fdf84625c411acb72fa51e12f9a8b067:
New code should use more modern `constexpr std::byte`s (and optionally `reserve()` vector):
```diff
 BOOST_AUTO_TEST_CASE(asmap_test_vectors)
 {
     // Randomly generated encoded ASMap with 128 ranges, up to 20-bit AS numbers.
-    static const auto ASMAP_DATA =
+    constexpr auto ASMAP_DATA{
         "fd38d50f7d5d665357f64bba6bfc190d6078a7e68e5d3ac032edf47f8b5755f87881bfd3633d9aa7c1fa279b3"
         "6fe26c63bbc9de44e0f04e5a382d8e1cddbe1c26653bc939d4327f287e8b4d1f8aff33176787cb0ff7cb28e3f"
         "daef0f8f47357f801c9f7ff7a99f7f9c9f99de7f3156ae00f23eb27a303bc486aa3ccc31ec19394c2f8a53ddd"
@@ -625,13 +625,14 @@ BOOST_AUTO_TEST_CASE(asmap_test_vectors)
         "1e3bbc17ba4d6f79ea3f031b876799ac268b1e0ea9babf0f9a8e5f6c55e363c6363df46afc696d7afceaf49b6"
         "e62df9e9dc27e70664cafe5c53df66dd0b8237678ada90e73f05ec60e6f6e96c3cbb1ea2f9dece115d5bdba10"
         "33e53662a7d72a29477b5beb35710591d3e23e5f0379baea62ffdee535bcdf879cbf69b88d7ea37c8015381cf"
-        "63dc33d28f757a4a5e15d6a08"_hex_v_u8;
+        "63dc33d28f757a4a5e15d6a08"_hex};
 
     // Convert to std::vector<bool> format that the ASMap interpreter uses.
     std::vector<bool> asmap_bits;
+    asmap_bits.reserve(ASMAP_DATA.size() * 8);
     for (auto byte : ASMAP_DATA) {
         for (int bit = 0; bit < 8; ++bit) {
-            asmap_bits.push_back((byte >> bit) & 1);
+            asmap_bits.push_back((std::to_integer<uint8_t>(byte) >> bit) & 1);
         }
     }

```

fjahr replied:
Done

hodlinator commented on src/test/netbase_tests.cpp:
```diff
@@ -610,4 +611,55 @@ BOOST_AUTO_TEST_CASE(isbadport)
+        "1e3bbc17ba4d6f79ea3f031b876799ac268b1e0ea9babf0f9a8e5f6c55e363c6363df46afc696d7afceaf49b6"
+        "e62df9e9dc27e70664cafe5c53df66dd0b8237678ada90e73f05ec60e6f6e96c3cbb1ea2f9dece115d5bdba10"
+        "33e53662a7d72a29477b5beb35710591d3e23e5f0379baea62ffdee535bcdf879cbf69b88d7ea37c8015381cf"
+        "63dc33d28f757a4a5e15d6a08"_hex_v_u8;
```
Review TODO - should I understand format, use tool to decode it?

hodlinator replied:
Tried:
```shell
₿ printf fd38d50f7d5d665357f64bba6bfc190d6078a7e68e5d3ac032edf47f8b5755f87881bfd3633d9aa7c1fa279b36fe26c63bbc9de44e0f04e5a382d8e1cddbe1c26653bc939d4327f287e8b4d1f8aff33176787cb0ff7cb28e3fdaef0f8f47357f801c9f7ff7a99f7f9c9f99de7f3156ae00f23eb27a303bc486aa3ccc31ec19394c2f8a53dddea3cc56257f3b7e9b1f488be9c1137db823759aa4e071eef2e984aaf97b52d5f88d0f373dd190fe45e06efef1df7278be680a73a74c76db4dd910f1d30752c57fe2bc9f079f1a1e1b036c2a69219f11c5e11980a3fa51f4f82d36373de73b1863a8c27e36ae0e4f705be3d76ecff038a75bc0f92ba7e7f6f4080f1c47c34d095367ecf4406c1e3bbc17ba4d6f79ea3f031b876799ac268b1e0ea9babf0f9a8e5f6c55e363c6363df46afc696d7afceaf49b6e62df9e9dc27e70664cafe5c53df66dd0b8237678ada90e73f05ec60e6f6e96c3cbb1ea2f9dece115d5bdba1033e53662a7d72a29477b5beb35710591d3e23e5f0379baea62ffdee535bcdf879cbf69b88d7ea37c8015381cf63dc33d28f757a4a5e15d6a08 \
| xxd -r -p | ./contrib/asmap/asmap-tool.py decode
::/7 AS961340
200::/7 AS693761
400::/6 AS379835
900::/8 AS231066
c00::/6 AS216064
1000::/9 AS66046
1080::/9 AS601879
1200::/7 AS672176
1400::/8 AS943262
1500::/8 AS902718
1600::/7 AS52631
1c00::/6 AS499880
2000::/7 AS916337
2400::/7 AS17255
2600::/7 AS48586
2800::/11 AS618608
2820::/11 AS80370
2840::/12 AS615085
2850::/12 AS585696
2860::/11 AS162102
2880::/10 AS494024
28c0::/11 AS982192
28e0::/12 AS610253
28f0::/12 AS1015065
2900::/8 AS527255
2b00::/9 AS393256
2c00::/12 AS279418
2c10::/14 AS146709
2c18::/13 AS317296
2c20::/11 AS576731
2c40::/12 AS943095
2c50::/13 AS100713
2c58::/13 AS522520
2c60::/12 AS461149
2c80::/13 AS15272
2c88::/13 AS297645
2ca0::/13 AS433935
2ca8::/13 AS633894
2cb0::/12 AS987819
2cc0::/11 AS770287
2ce0::/12 AS747699
2cf0::/12 AS622442
2d00::/9 AS464074
2d84::/14 AS1034916
2d88::/14 AS637424
2d8c::/15 AS602497
2d8e::/15 AS1995
2d90::/12 AS202022
2da0::/13 AS688185
2da8::/13 AS148562
2db0::/17 AS444111
2db0:8000::/17 AS728945
2db2::/15 AS123659
2db8::/15 AS626395
2dba::/15 AS340438
2dbc::/16 AS609886
2dbd::/16 AS417213
2dc0::/12 AS16298
2dd0::/12 AS735128
2de0::/12 AS506083
2df0::/13 AS436493
2e40::/10 AS500584
2e80::/11 AS991607
2ee0::/11 AS925272
2f00::/9 AS707739
2f80::/9 AS328114
4000::/5 AS248495
5000::/7 AS124471
5200::/7 AS539993
5400::/6 AS374443
5800::/5 AS999782
6000::/5 AS407935
6800::/5 AS435070
7000::/4 AS677580
8000::/4 AS244121
9000::/4 AS862116
a000::/9 AS863404
a100::/9 AS245226
a180::/11 AS915050
a1c0::/11 AS811862
a400::/7 AS824696
a600::/7 AS384156
a800::/7 AS150135
aa40::/10 AS862795
aa80::/9 AS276407
ac20::/11 AS171508
ac40::/11 AS22548
ac60::/11 AS405134
ac80::/10 AS1011589
acc0::/10 AS998011
ad00::/9 AS451326
ad80::/9 AS16307
ae00::/8 AS1022594
af00::/9 AS257981
b000::/5 AS475471
c000::/3 AS969411
e000::/3 AS824019
```

hodlinator commented on src/test/streams_tests.cpp:
```diff
@@ -111,6 +133,7 @@ BOOST_AUTO_TEST_CASE(xor_file)
 #endif
         AutoFile xor_file{raw_file(mode), obfuscation};
         xor_file << test1 << test2;
+        BOOST_CHECK_EQUAL(xor_file.size(), 14);
```
14 is from something like 2 * 4-byte vector sizes + 2+3 bytes data + EOF? Would expect smaller compact sizes. 🤔 

Tried adding...
```C++
    {
        AutoFile test{raw_file("rb"), obfuscation};
        BOOST_CHECK_EQUAL(test.size(), 14);
    }
```
...just after this block and then start digging into the data but it fails with size suddenly being 7?

hodlinator commented on src/wallet/migrate.cpp:
```diff
@@ -544,8 +544,7 @@ void BerkeleyRODatabase::Open()
     // Verify the size of the file is a multiple of the page size
-    db_file.seek(0, SEEK_END);
-    int64_t size = db_file.tell();
+    int64_t size{db_file.size()};
```
nit:
```suggestion
    const int64_t size{db_file.size()};
```

hodlinator reviewed (commented):
Reviewed 5bb456c92bbd27ebbba1773832b051968f162b8a

Thanks for providing a less intimidating PR as a stepping stone for reviewers new to this domain like myself.

---

PR description still mentions 4 commits but only 3 are currently present (after https://github.com/bitcoin/bitcoin/pull/33026#discussion_r2429247548).

fjahr committed:
test: Add better coverage for Autofile size()

The new test explicitly checks that the function does not change the current position.

hodlinator reviewed (approved):
tACK 7f318e1dd0496384e7bc6d8754c5a2a618a14b2a

Happy to have helped uncover a bug (https://github.com/bitcoin/bitcoin/pull/33026#discussion_r2527777820). Thanks for incorporating my suggestions! New commit with `size_preserves_position`-test looks good, I especially like how it verifies that one can ask for size at the end but not read without triggering an exception.

Started looking at #28792 and agree this is a reasonable subset of changes to peel off (although I would also consider peeling off bits->bytes refactor and span-usage to its own PR).

laanwj reviewed (approved):
Code review ACK 7f318e1dd0496384e7bc6d8754c5a2a618a14b2a
These are some straightforward changes that even make sense outside the context of embedded ASmap.

waketraindev commented:
I think the `AutoFile::size` implementation and the subsequent refactoring would fit better in their own dedicated refactoring PR, or in a PR where they're actually required. They don't seem to be needed by either of the other two commits (the `LogWarning` change in `asmap.cpp` or the newly added test vectors in `netbase_tests.cpp`).

I understand this is listed as "selected minor preparatory work", and I realize this is nitpicking a bit, but it would still be cleaner to introduce these changes in a PR where they're actually used or needed, or in a dedicated Streams refactoring PR. For that matter, they already appear to be included in #33878 as well as in #28792. The need for this refactor in the current asmap change set also seems fairly low, since it only removes ~3 lines of `seek` & `tell` from `asmap.cpp` while touching files that aren't directly related to the asmap work here (e.g., `src/wallet/migrate.cpp`, `src/streams.{cpp,h}`).

Just to double-check, I tested the `netbase_tests.cpp` changes with the `AutoFile::size` commits removed, and everything still passed:

```
*** No errors detected
```

waketraindev commented on src/test/netbase_tests.cpp:
```diff
@@ -610,4 +611,56 @@ BOOST_AUTO_TEST_CASE(isbadport)
+    for (auto byte : ASMAP_DATA) {
+        for (int bit = 0; bit < 8; ++bit) {
+            asmap_bits.push_back((std::to_integer<uint8_t>(byte) >> bit) & 1);
+        }
```
> The call should be fully optimized out anyway, no?

I sure hope so!
It's test-only code, so not a big deal.

fjahr commented:
> I think the `AutoFile::size` implementation and the subsequent refactoring would fit better in their own dedicated refactoring PR, or in a PR where they're actually required. They don't seem to be needed by either of the other two commits (the `LogWarning` change in `asmap.cpp` or the newly added test vectors in `netbase_tests.cpp`).

Most of our refactoring commits are not "actually required", strictly speaking. We still do them for a variety of reasons. In this case, I was working on the asmap implementation code for a while and this piece of it always jumped out at me as a layer violation. It seemed like this kind of file handling functionality should live in a helper function in a different part of the code base. So that's what I did. I probably still hadn't done it if there had been no deduplication opportunity at all, but at least there was one. And an additional benefit is having the function also allows for testing of this functionality in isolation.

> I understand this is listed as "selected minor preparatory work", and I realize this is nitpicking a bit, but it would still be cleaner to introduce these changes in a PR where they're actually used or needed, or in a dedicated Streams refactoring PR. For that matter, they already appear to be included in #33878 as well as in #28792. The need for this refactor in the current asmap change set also seems fairly low, since it only removes ~3 lines of `seek` & `tell` from `asmap.cpp` while touching files that aren't directly related to the asmap work here (e.g., `src/wallet/migrate.cpp`, `src/streams.{cpp,h}`).

The code is used in asmap and the major benefit IMO is that there is no code that doesn't really belong in the asmap implementation distracting from the actual meat there. While the impact might be minor it can still help because a lot of people (myself included) have struggled with the review of this code. And direct usage is the usual requirement we have for including refactorings in a feature PR afaict unless things become too big or a separate conceptual discussion is required. I think the extend of this is fine in the in terms of LOC and the other files it touches. So I will not move this out because practically speaking there are now two active and three stale acks here plus there has been prior review of this in #28792. Pulling this out, looking for another PR to include this in or putting it into it's own PR would create too much friction and distract more from moving asmap forward than keeping it here. It's true that we generally want smaller, focussed PRs but we also shouldn't overdo this and in this case I fall on having these changes bundled here is still the best way to go.

maflcko commented on src/test/streams_tests.cpp:
```diff
@@ -781,4 +807,41 @@ BOOST_AUTO_TEST_CASE(streams_hashed)
+
+    // Case: Pos at EOF
+    f.seek(0, SEEK_END);
+    (void)f.size();
```
nit: Here and above in the last commit: Why not check the size is equal to 10 each time?

maflcko reviewed (commented):
forgot to send the nits, basically just a test nit to check the returned size, but just a nit
